### PR TITLE
Library Forwarding: Support libGL/libvulkan without forwarding libX11

### PR DESCRIPTION
**Pull request by neobrain**
~~*Marked as draft because this will need **a lot** of testing*~~

## Overview

Currently when forwarding libGL/libvulkan to host libraries, FEX generally requires libX11 calls to be forwarded to the aarch64 host as well to do anything interesting. This brings a number of problems, since the libX11 API uses complex data structures that are difficult to adapt for an aarch64 host without breaking the x86 client application. Most importantly, forwarding libX11 from a 32-bit application to the 64-bit host in the current form is impossible since the layout of these shared data structures will be inconsistent.

This PR removes FEX's current libX11 forwarding support and instead makes the host-side libGL/libvulkan interoperable with an X11 connection managed by the guest. This idea was initially brought up in #1735, but the implementation was reliant on adding tremendous amounts of boilerplate.

## Implementation

Interoperability is enabled by having our host-side libGL/libvulkan wrappers open a shadow connection to the X server (either via `xcb_connect` or `XOpenDisplay`). Since most libX11 objects are global identifiers (`XID`s) that are valid in any context, the host-side drivers can talk over this shadow connection as if it were the same as the guest-side connection.

This means we can pass XIDs across architecture boundaries without modification. Only the connection types (`xcb_connection_t` or `_XDisplay`) must be exchanged for the shadow connection (and vice-versa) when crossing architectures.

Furthermore, `XVisualInfo` is an exceptional libX11 type since it's neither a type alias for `XID` nor a connection. These objects can't directly be shared between guest and host due to the embedded `Visual*` object. However, we can use `XGetVisualInfo` to look up a valid `XVisualInfo` whenever we cross architectures.

Finally, some libX11 functions used by libGL allocate data internally and require the user to de-allocate it using XFree. Since XFree is a macro that expands to `free`, data allocated by the host-side GL driver (on the host heap) would be de-allocated as if it were on the guest heap. To handle this properly, FEX wraps data-allocating libGL functions and relocates their returned data to the guest stack.

## Impact

This PR has the potential to fix any game that could previously not use libGL/libvulkan forwarding due to issues in libX11.

This PR also has the potential to break any games that use the glX/WSI APIs in an unexpected way.

## Testing

I have thrown this PR at all games I can get running in my Linux VM with virgl / llvmpipe and could not find any regressions. Unless otherwise noted, these all work:
* 7 Billion Humans ✅
* (Crypt of the Necrodancer: Had issues with `glPushDebugGroup` specifically but otherwise worked fine; needs a re-test) ☑️ (previously fully broken, see #2173)
* Dandara: Trials of Fear Edition ✅
* Ender Lilies ✅
* Enter the Gungeon ✅
* FTL ✅
* Jupiter Hell ✅
* Just Shapes and Beats ✅
* Magicka 2 ✅
* SteamWorld Dig 2 ✅
* Super Hexagon ✅
* Super Meat Boy ✅
* VVVVVV ✅ (previously broken, see #3481)
* World of Goo ✅
* Xonotic (both GLX/SDL builds) ✅

Other notable content:
* eglgears: Fails to present rendered images. This demo is actually doing surprisingly complex low-level operations, of which I'm not sure how well-behaved they are.
* glxgears ✅
* MangoHud ✅
* Steam (with steamwebhelper AppConfig, see below) ✅
* steamwebhelper: This circumvents `glX` and instead manually constructs GLX protocol requests for `libxcb-glx`. libGL forwarding is hence forcefully disabled for this process in #3580.
* zink: Seems to work for some games but not others. (Needs further testing to see if this is a regression or just a limitation of lavapipe) ☑️

It's worth noting that these promising test results also extend to 32-bit libGL support, which will be submitted in a follow-up PR.

## TODO

- [x] Verify Wayland support still works
- [x] Lots of testing
- [x] Merge #3580 and #3581

### CODE REVIEW

**Comment by neobrain** [comment]
Ender Lilies regressed at some point during development, so it's probably best to focus testing on GL titles for the time being.

**Comment by neobrain** [comment]
More games tested on X13s:
* Celeste ✅
* FEZ ✅
* Filament (starts but seems to be stuck on loading intro?)
* Hollow Knight (starts but seems to be stuck on loading intro?)
* Magicka 2 ✅
* The Lightbringer ✅
* The Binding of Isaac: Rebirth ✅
* Unbound: Worlds Apart (starts but seems to be stuck on loading intro?)
* Yooka Laylee ✅

I think these regressions are all Vulkan-based, so that's something to look into. There's one WSI function that hasn't been implemented yet, which might be the culprit.

**Comment by neobrain** [comment]
Added a patch to also adapt `vkCreateXlibSurfaceKHR` for this new approach. The code should probably be shared with libGL instead, but at least this gets rid of the few known regressions for testing.

**Comment by Sonicadvance1** [comment]
Initial testing results. I still need to double check that the notes on various games are not regressions from this PR.
```
# Proton
Works:
Animal Super Squad
Batman - The Telltale series
Bravely Default 2
Claybook
Crash Bandicoot 4: It's about Time
Crash Bandicoot N. Sane Trilogy
The Dark Pictures Anthology: Little Hope
Dark Souls: Remastered
Doom 1993
The Elder Scrolls V: Skyrim Special Edition
Exit the Gungeon
Final Fantasy
Final Fantasy 2
Final Fantasy 3
Final Fantasy 4
Final Fantasy 5
Final Fantasy 7 Remake intergrade
Final Fantasy 12
Ghostrunner
God of War 2018
Hades
Halo: The Master Chief Collection
Horizon Zero Dawn
INSIDE
Jusant
Kamiko
Legend of Mana
Life is Strange: True Colors
Marvel's Spider-Man: Miles Morales
  - Thunks worked until Orin exploded? Needs retest.
Marvel's Spider-Man Remastered
Metal Gear Solid 3: Snake Eater - Master Collection Version
Monster Hunter World
Nier Replicant
Nier: Automata
Okami HD
The Outer Worlds
Resident Evil 2
Resident Evil 3
River City Girls
The Room 4: Old Sins
Shantae: Half-Genie Hero Ultimate Edition
Stray
Tetris Effect: Connected
World of Final Fantasy

Doesn't work:
Chrono Chross: The Radical Dreamers Edition
  - Complains about insufficient vram
Crysis Core - Final Fantasy 7 - Reunion
  - Infinite looping on futex
Deliver Us The Moon
  - Exploded the Orin board so hard that it reboot
Detroit: Become Human
  - Issue with shader caching causing norendering?
Devil May Cry 5
  - Infinite looping on futex
Devolverland Expo
  - Exploded the Orin board so hard that it reboot
Dirt Rally 2.0
  - Infinite looping on futex
Doom 2016
  - Infinite looping on futex
Ghosts 'n Goblins Ressurrection
  - Infinite looping on futex
  - Didn't even work without thunks
The Great Ace Attoney Chronicles
  - Didn't even work without thunks
Guilty Gear Strive
  - Infinite looping on futex
Hitman 2
  - Infinite looping on futex
Hitman World of Assassination
  - Infinite looping on futex
Injustice 2
  - Didn't even work without thunks
Mortal Kombat 11
  - Black-screen even without thunks
OneShot
  - Hangs even without thunks
Resident Evil 5
  - Enigma DRM was added since last test
Resident Evil 7
  - Infinite looping on futex
Resident Evil Village
  - Crashes even without thunks
SOMA
  - Infinite looping on futex
Teardown
  - Infinite looping on futex
Thumper
  - Infinite looping on futex
Wolfenstein II: The New Colossus
  - Infinite looping on futex

32-bit:
20XX
Alan Wake American Nightmare
Chrono Trigger
Crayon Physics Deluxe
Danganronpa 2: Goodbye Despair
Dead Space 2008
Divinity: Original Sin Classic
Fallout: New Vegas
Final Fantasy 7
  - Intro FMVs magenta for some reason
  - RENDERING BUGS!
Final Fantasy 8 remastered
Gish
Half Minute Hero: Super Mega Neo Climax Ultimate Boy
The Henry Stickmin Collection
Left 4 Dead
Metal Gear Rising: Revengeance
Mighty Switch Force! Collection
Mighty Switch Force! Hose It Down!
Mighty Switch Force! Hyper Drive Edition
No More Heroes
Noita
Orcs Must Die!
Phoenix Wright: Ace Attorney Trilogy
Pinball Wicked
Sonic Mania
Police Stories
The Room
Secret of Mana
Shantae and the Pirate's Curse
Shantae: Risky's Revenge - Director's Cut
Sonic Lost World
Spelunky
Tales of Symphonia
The Wonderful 101: Remastered

# Linux
Works:
7 Billion Humans
Aperture Desk Job
Dota Underlords
Dwarf Fortress
Ender Lilies
Filament
  - Has some rendering defects on title logo
Unbound Worlds Apart

Doesn't Work:
Beyond a Steel Sky
  - "Cannot find a compatible Vulkan driver (ICD)"
Firewatch
  - SIGABRT without thunks
  - With Thunks
  - glXGetProcAddress: not found glFramebufferTexture2DMultisampleEXT
  - glXGetProcAddress: not found glFramebufferTexture2DMultisampleIMG
  - glXGetProcAddress: not found glResolveMultisampleFramebufferAPPLE
  - Then close or crash
Psychonauts 2
  - Hanging with black screen after logos both with and without thunks?

32-Bit:

# Misc
Larian Launcher bugged again?
Applications having a chance of hanging on shutdown?
  - Dota Underlords
  - Unbound Worlds apart
Applications have a chance of crashing on shutdown?
  - Unbound Worlds apart

```

**Comment by Sonicadvance1** [comment]
I had to revert the ThunksDB dependencies in order to work around the infinite looping futex bug with this PR.
Backtrace puts this in the following backtrace.
```
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".
__GI___lll_lock_wake (futex=futex@entry=0x7fffdee55aa8, private=private@entry=0) at ./nptl/lowlevellock.c:64
64      ./nptl/lowlevellock.c: No such file or directory.
(gdb) bt
#0  __GI___lll_lock_wake (futex=futex@entry=0x7fffdee55aa8, private=private@entry=0) at ./nptl/lowlevellock.c:64
#1  0x0000ffffb70d20f0 in lll_mutex_unlock_optimized (mutex=0x7fffdee55aa8) at ./nptl/pthread_mutex_unlock.c:43
#2  __GI___pthread_mutex_unlock_usercnt (mutex=mutex@entry=0x7fffdee55aa8, decr=decr@entry=0) at ./nptl/pthread_mutex_unlock.c:68
#3  0x0000ffffb70cc760 in __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x7fffdee55aa8, cond=0x7fffdee56c00) at ./nptl/pthread_cond_wait.c:405
#4  ___pthread_cond_wait (cond=0x7fffdee56c00, mutex=0x7fffdee55aa8) at ./nptl/pthread_cond_wait.c:627
#5  0x00007ffff775b730 in ?? () from target:/lib/aarch64-linux-gnu/libxcb.so.1
#6  0x00007ffff7762cf8 in xcb_send_request_with_fds64 () from target:/lib/aarch64-linux-gnu/libxcb.so.1
#7  0x00007ffff776326c in xcb_send_request () from target:/lib/aarch64-linux-gnu/libxcb.so.1
#8  0x00007ffff776b37c in xcb_query_extension () from target:/lib/aarch64-linux-gnu/libxcb.so.1
#9  0x00007ffff7762880 in ?? () from target:/lib/aarch64-linux-gnu/libxcb.so.1
#10 0x00007ffff7762990 in xcb_get_extension_data () from target:/lib/aarch64-linux-gnu/libxcb.so.1
#11 0x00007ffff7762bd4 in xcb_send_request_with_fds64 () from target:/lib/aarch64-linux-gnu/libxcb.so.1
#12 0x00007ffff776326c in xcb_send_request () from target:/lib/aarch64-linux-gnu/libxcb.so.1
#13 0x00007fffdb996ae0 in xcb_randr_query_version () from target:/lib/aarch64-linux-gnu/libxcb-randr.so.0
#14 0x00007fff9c56517c in wsi_display_output_to_root () from target:/usr/lib/aarch64-linux-gnu/libvulkan_radeon.so
#15 0x00007fff9c5652dc in wsi_GetRandROutputDisplayEXT () from target:/usr/lib/aarch64-linux-gnu/libvulkan_radeon.so
#16 0x00007fffdbfc76f8 in GuestWrapperForHostFunction<VkResult (VkPhysicalDevice_T*, _XDisplay*, unsigned long, unsigned long*), VkPhysicalDevice_T*, _XDisplay*, unsigned long, unsigned long*>::Call<ParameterAnnotations{}, ParameterAnnotations{false, true}, ParameterAnnotations{false, true}, ParameterAnnotations{}, ParameterAnnotations{}>(void*)::{lambda(guest_layout<VkPhysicalDevice_T*>, guest_layout<_XDisplay*>, guest_layout<unsigned long>, guest_layout<unsigned long*>, unsigned long)#1}::operator()(guest_layout<VkPhysicalDevice_T*>, guest_layout<_XDisplay*>, guest_layout<unsigned long>, guest_layout<unsigned long*>, unsigned long) const (args=..., args=..., args=..., args=..., target=140736932960952, this=<optimized out>)
    at /mnt/Work/Work/work/FEXNew/ThunkLibs/HostLibs/../include/common/Host.h:612
#17 Invoke<guest_layout<VkResult>, guest_layout<VkPhysicalDevice_T*>, guest_layout<_XDisplay*>, guest_layout<unsigned long>, guest_layout<unsigned long*>, unsigned long, GuestWrapperForHostFunction<VkResult (VkPhysicalDevice_T*, _XDisplay*, unsigned long, unsigned long*), VkPhysicalDevice_T*, _XDisplay*, unsigned long, unsigned long*>::Call<ParameterAnnotations{}, ParameterAnnotations{false, true}, ParameterAnnotations{false, true}, ParameterAnnotations{}, ParameterAnnotations{}>(void*)::{lambda(guest_layout<VkPhysicalDevice_T*>, guest_layout<_XDisplay*>, guest_layout<unsigned long>, guest_layout<unsigned long*>, unsigned long)#1}&>(GuestWrapperForHostFunction<VkResult (VkPhysicalDevice_T*, _XDisplay*, unsigned long, unsigned long*), VkPhysicalDevice_T*, _XDisplay*, unsigned long, unsigned long*>::Call<ParameterAnnotations{}, ParameterAnnotations{false, true}, ParameterAnnotations{false, true}, ParameterAnnotations{}, ParameterAnnotations{}>(void*)::{lambda(guest_layout<VkPhysicalDevice_T*>, guest_layout<_XDisplay*>, guest_layout<unsigned long>, guest_layout<unsigned long*>, unsigned long)#1}&, PackedArguments<guest_layout<VkResult>, guest_layout<VkPhysicalDevice_T*>, guest_layout<_XDisplay*>, guest_layout<unsigned long>, guest_layout<unsigned long*>, unsigned long>&) (args=..., func=...)
    at /mnt/Work/Work/work/FEXNew/ThunkLibs/HostLibs/../include/common/PackedArguments.h:547
#18 GuestWrapperForHostFunction<VkResult (VkPhysicalDevice_T*, _XDisplay*, unsigned long, unsigned long*), VkPhysicalDevice_T*, _XDisplay*, unsigned long, unsigned long*>::Call<ParameterAnnotations{}, ParameterAnnotations{false, true}, ParameterAnnotations{false, true}, ParameterAnnotations{}, ParameterAnnotations{}>(void*) (argsv=0x1000fc7f8)
    at /mnt/Work/Work/work/FEXNew/ThunkLibs/HostLibs/../include/common/Host.h:617
#19 0x00007fffe36a6e14 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
(gdb)```

**Comment by neobrain** [comment]
Thanks for testing!

I added the `_XDisplay`/`xcb_connection_t` hook to a few more Vulkan functions that needed it, and I restructed the existing code to ensure we caught them all this time. Since your backtrace points to `vkGetRandROutputDisplayEXT`, this should fix that particular crash and hopefully gets Thumper working. Could you retest?

**Comment by neobrain** [comment]
(CI is failing in unit tests due to non-severe side-effects that don't affect game testing)

**Comment by Sonicadvance1** [comment]
Same as with my hack, the games that were infinite looping are permanently hanging on a futex.

**Comment by Sonicadvance1** [comment]
Looks like this is wine hanging in `pthread_mutex_lock`, waiting for an internal mutex to unlock and never does. Happening at the following call site.
https://github.com/wine-mirror/wine/blob/master/dlls/winex11.drv/x11drv_main.c#L242

**Comment by neobrain** [comment]
I managed to reproduce the hangs in Thumper locally and found an additional issue with `vkGetRandROutputDisplayEXT`: The first time the application/proton creates a Vulkan instance, it doesn't request `VK_EXT_acquire_xlib_display`, so the host driver returns `nullptr` for `vkGetRandROutputDisplayEXT`. Since we cache these results, we never initialized `fexldr_ptr_libvulkan_vkGetRandROutputDisplayEXT` properly.

I added a simple workaround to force re-loading this particular function if it's still null. Locally, that fixes Thumper and hopefully the other hangs you were observing.

> Backtrace puts this in the following backtrace. (Using Thumper since it consistently reproduces).

How did you create this backtrace? All my attempts of producing anything as clean failed, so for future reference it would be really useful to know how to handle Proton-backed games.

**Comment by neobrain** [comment]
> Now the only regression I've found is Psychonauts 2, it plays all of its intro logos and then hangs before showing the title screen. While it works on main.

I can't reproduce that issue. Tried on X13s in an i3 X11 session with GL and Vulkan enabled in thunks.json. Is there anything else specific about your setup? Could you try and see if changing window managers and/or testing on (X)Wayland might help?

**Comment by Sonicadvance1** [comment]
Interesting, tried on the X13s and it got to the title screen there.
So this looks like it is an Orin+Radeon specific quirk. Maybe using some specific vulkan extensions or something. Can probably be a side investigation

**Comment by Sonicadvance1** [comment]
Upgraded my Orin board to Ubuntu 24.04 and the Psychonauts 2 hang went away. So that's a thing I guess.

**Comment by neobrain** [comment]
Squashed the fixup patches in the branch and merged the duplicated X11Manager from GL/Vulkan into a single header.

> Interesting, tried on the X13s and it got to the title screen there.
> So this looks like it is an Orin+Radeon specific quirk. Maybe using some specific vulkan extensions or something. Can > probably be a side investigation

> Upgraded my Orin board to Ubuntu 24.04 and the Psychonauts 2 hang went away. So that's a thing I guess.

Good to know this only happens on older Ubuntu versions (22.04). Sadly all my attempts at setting up a environment older than 24.04 have led nowhere, so I couldn't try to debug the problem.

Since the changes are running well otherwise *and* since the game appears to work fine on recent distributions, I'd propose merging the PR as-is.

File: ThunkLibs/libvulkan/Host.cpp [hunk 1]
```diff
@@ -172,6 +240,25 @@ static PFN_vkVoidFunction FEXFN_IMPL(vkGetInstanceProcAddr)(VkInstance a_0, cons
...
+    // VkInstance with no extensions before creating its actual instance.
+    using namespace std::string_view_literals;
+    if (a_1 == "vkGetRandROutputDisplayEXT"sv && !LDR_PTR(vkGetRandROutputDisplayEXT)) {
+      (void*&)LDR_PTR(vkGetRandROutputDisplayEXT) = (void*)LDR_PTR(vkGetInstanceProcAddr)(a_0, "vkGetRandROutputDisplayEXT");
+    }
+    if (a_1 == "vkAcquireXlibDisplayEXT"sv && !LDR_PTR(vkAcquireXlibDisplayEXT)) {
+      (void*&)LDR_PTR(vkAcquireXlibDisplayEXT) = (void*)LDR_PTR(vkGetInstanceProcAddr)(a_0, "vkAcquireXlibDisplayEXT");
+    }
+    if (a_1 == "vkGetPhysicalDeviceXcbPresentationSupportKHR"sv && !LDR_PTR(vkGetPhysicalDeviceXcbPresentationSupportKHR)) {
+      (void*&)LDR_PTR(vkGetPhysicalDeviceXcbPresentationSupportKHR) = (void*)LDR_PTR(vkGetInstanceProcAddr)(a_0, "vkGetPhysicalDeviceXcbPre"
```
Comment (Sonicadvance1) [inline ThunkLibs/libvulkan/Host.cpp:254]: This weird string break should be reformatted to not look terrible. Same with the one in the next if statement.

Comment (neobrain) [reply ThunkLibs/libvulkan/Host.cpp:254]: I've been trying a couple of variations on that line and couldn't get `clang-format` to produce a different result. We could turn off formatting locally, but not sure the two-line preprocessor noise added by that is any better.

Happy to proceed with whichever route you prefer.

Comment (neobrain) [reply ThunkLibs/libvulkan/Host.cpp]: Ah yes, hadn't tried a local variable. Definitely less disruptive than the options I considered :)

**Review by Sonicadvance1 — APPROVED** [review]

**Pull request merged**